src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs

```diff
@@ -135,33 +135,46 @@ public RegexCode RegexCodeFromRegexTree(RegexTree tree)
             bool rtl = (tree.Options & RegexOptions.RightToLeft) != 0;
 
             // Compute prefixes to help optimize FindFirstChar.
-            RegexBoyerMoore? bmPrefix = null;
-            RegexPrefix? fcPrefix = null;
-            RegexPrefix prefix = RegexFCD.Prefix(tree);
-            if (prefix.Prefix.Length > 1 && prefix.Prefix.Length <= RegexBoyerMoore.MaxLimit) // if it's <= 1 || > MaxLimit, perf is better using fcPrefix
+            RegexBoyerMoore? boyerMoorePrefix = null;
+            RegexPrefix[]? leadingCharClasses = null;
+            RegexPrefix leadingSubstring = RegexPrefixAnalyzer.ComputeLeadingSubstring(tree);
+            if (leadingSubstring.Value.Length > 1 && // if it's <= 1, perf is better using leadingCharClasses
+                leadingSubstring.Value.Length <= RegexBoyerMoore.MaxLimit)
             {
                 // Compute a Boyer-Moore prefix if we find a single string of sufficient length that always begins the expression.
                 CultureInfo culture = (tree.Options & RegexOptions.CultureInvariant) != 0 ? CultureInfo.InvariantCulture : CultureInfo.CurrentCulture;
-                bmPrefix = new RegexBoyerMoore(prefix.Prefix, prefix.CaseInsensitive, rtl, culture);
+                boyerMoorePrefix = new RegexBoyerMoore(leadingSubstring.Value, leadingSubstring.CaseInsensitive, rtl, culture);
             }
             else
             {
-                // If we didn't find such a string, try to compute the characters set that might begin the string.
-                fcPrefix = RegexFCD.FirstChars(tree);
+                // If we didn't find a single leading substring, try to compute the characters set that might begin the string.
+                // First we employ a less aggressive but more valuable computation to see if we can find sets for each of the first N
+                // characters in the string.  If that's unsuccessful, we employ a more aggressive check to compute a set for just
+                // the first character in the string.
+
+                if ((tree.Options & RegexOptions.Compiled) != 0) // currently not utilized by the interpreter
+                {
+                    leadingCharClasses = RegexPrefixAnalyzer.ComputeMultipleCharClasses(tree, maxChars: 5); // limit of 5 is based on experimentation and can be tweaked as needed
+                }
+
+                if (leadingCharClasses is null)
+                {
+                    RegexPrefixAnalyzer.ComputeFirstCharClass(tree);
```

Comment:
should this be assigning the return value?

Reply:
Ugh, yes, I must have somehow deleted the assignment just before hitting submit. Thank you for catching that.

Reply:
Did it break any test?

Reply:
> Did it break any test?

No, it's a perf optimization.  It would impact perf tests, but nothing functional would catch it.